src/libs/actions/OnyxUpdateManager.ts

```diff
@@ -27,6 +28,132 @@ Onyx.connect({
+let deferredUpdateBeforeReconnect: OnyxUpdatesFromServer | undefined;
+let deferredUpdates: DeferredUpdatesDictionary = {};
+
+// This function will unpause the SequentialQueue, log an info and reset the deferred updates
+function unpauseQueueAndReset() {
+    console.debug('[OnyxUpdateManager] Done applying all updates');
+    deferredUpdates = {};
+    Onyx.set(ONYXKEYS.ONYX_UPDATES_FROM_SERVER, null);
+    SequentialQueue.unpause();
+}
+
+// This function applies a list of updates to Onyx in order and resolves when all updates have been applied
+const applyUpdates = (updates: DeferredUpdatesDictionary) => Promise.all(Object.values(updates).map((update) => OnyxUpdates.apply(update)));
+
+type DetectGapAndSplitResult = {applicableUpdates: DeferredUpdatesDictionary; updatesAfterGaps: DeferredUpdatesDictionary; latestMissingUpdateID: number | undefined};
+
+// In order for the deferred updates to be applied correctly in order,
+// we need to check if there are any gaps deferred updates.
+function detectGapsAndSplit(updates: DeferredUpdatesDictionary): DetectGapAndSplitResult {
+    const updateValues = Object.values(updates);
+    const applicableUpdates: DeferredUpdatesDictionary = {};
+
+    let gapExists = false;
+    let firstUpdateAfterGaps: number | undefined;
+
+    for (const [index, update] of updateValues.entries()) {
+        // If it's the first update, we need to check that the previousUpdateID of the fetched update is the same as the lastUpdateIDAppliedToClient
+        // If any update's previousUpdateID doesn't match the lastUpdateID from the previous update, there's a gap in the deferred updates
+        const isFirst = index === 0;
+        if (isFirst && update.previousUpdateID !== lastUpdateIDAppliedToClient) {
+            break;
+        }
+
+        if (updates[Number(update.previousUpdateID)]) {
+            // If a gap exists, we will not add any more updates to the applicable updates
+            // Instead, we we have to keep track of the first update after all detected gaps
+            if (gapExists) {
+                // If "firstUpdateAfterGaps" hasn't been set yet and there was a gap,
+                // we need to set it to the first update after all gaps
+                if (!firstUpdateAfterGaps) {
+                    firstUpdateAfterGaps = Number(update.previousUpdateID);
+                }
+            } else {
+                // If no gap exists yet and the previousUpdateID of the current update is found in the deferred updates, we can add it to the applicable updates
+                applicableUpdates[Number(update.lastUpdateID)] = update;
+            }
+        } else {
+            // When we find a(nother) gap, we need to set the flag to true and reset the "firstUpdateAfterGaps" variable, so that we can continue searching for the next update after all gaps
+            gapExists = true;
+            firstUpdateAfterGaps = undefined;
+        }
```

Comment:
It took me a while to understand what is happening here, can you please add more comments explaining?

Reply:
i simplified the conditions a bit, which removes unnecessary `if else` blocks. The reason we don't have early-returns here, is because `continue` is forbidden. This would make it more clean, but i think this is almost perfect